### PR TITLE
Fix menu vote handler to work with direct arrays.

### PR DESCRIPTION
**Pull request by dvander**

### CODE REVIEW

File: core/logic/smn_menus.cpp [hunk 1]
```diff
@@ -496,99 +492,61 @@ void CMenuHandler::OnMenuVoteResults(IBaseMenu *menu, const menu_vote_result_t *
...
+		if (!pContext->HeapAlloc2dArray(results->num_items, 2, &item_array_address, init.get())) {
+			g_DbgReporter.GenerateError(pContext, m_fnVoteResult, -1,
+										"Menu callback could not allocate %d cells for item list.",
+										item_array_size);
+			ok = false;
 		}
 	}
+
+	/* Finally, push everything */
+	if (!ok) {
```
Comment (PerfectLaugh) [inline core/logic/smn_menus.cpp:534]: This is not okay

File: core/logic/smn_menus.cpp [hunk 2]
```diff
@@ -496,99 +492,61 @@ void CMenuHandler::OnMenuVoteResults(IBaseMenu *menu, const menu_vote_result_t *
...
-			m_pVoteResults->PushCell(menu->GetHandle());
-			m_pVoteResults->PushCell(results->num_votes);
-			m_pVoteResults->PushCell(results->num_clients);
-			m_pVoteResults->PushCell(client_array_address);
-			m_pVoteResults->PushCell(results->num_items);
-			m_pVoteResults->PushCell(item_array_address);
-			m_pVoteResults->Execute(NULL);
+		if (!pContext->HeapAlloc2dArray(results->num_clients, 2, &client_array_address, init.get())) {
+			g_DbgReporter.GenerateError(pContext, m_fnVoteResult, -1,
+										"Menu callback could not allocate cells for client list.");
```
Comment (PerfectLaugh) [inline core/logic/smn_menus.cpp:512]: better to add %d into the message

**Review by PerfectLaugh — CHANGES_REQUESTED** [review]
Have to fix the sourcepawn with this PR https://github.com/alliedmodders/sourcepawn/pull/716

File: core/logic/smn_menus.cpp (hunk 2 above)
Comment (dvander) [reply core/logic/smn_menus.cpp:512]: The number is kind of meaningless. It's not accurate given that we allocate metadata behind the scenes, and at most it'll be a pretty small number.

**Comment by dvander** [comment]
Latest patch uses AutoEnterHeapScope instead of HeapPop, which is now basically deprecated.

**Review by asherkin — APPROVED** [review]
I assume this doesn't need the SP pin update any more?

**Review by Headline — APPROVED** [review]

**Pull request merged**